core/src/main/java/org/wildfly/plugins/bootablejar/BootableJarSupport.java

```diff
@@ -0,0 +1,309 @@
+            }
+            // Extract the cloud extension if cloud execution context.
+            if (bootableJARVersion != null) {
+                try (InputStream stream = BootableJarSupport.class.getClassLoader().getResourceAsStream("config.properties")) {
+                    Properties props = new Properties();
+                    props.load(stream);
+                    unzipCloudExtension(contentRootDir, bootableJARVersion, resolver);
+                    // Needed by extension
+                    Path marker = contentRootDir.resolve("openshift.properties");
+                    Files.createFile(marker);
+                }
+            }
+            buildJar(contentRootDir, targetJarFile, bootable, resolver);
+        } finally {
+            IoUtils.recursiveDelete(contentRootDir);
+        }
+    }
+
+    public static void unzipCloudExtension(Path contentDir, String version, MavenRepoManager resolver) throws MavenUniverseException, IOException {
+        MavenArtifact ma = new MavenArtifact();
+        ma.setGroupId("org.wildfly.plugins");
+        ma.setArtifactId("wildfly-jar-cloud-extension");
+        ma.setExtension("jar");
+        ma.setVersion(version);
+        resolver.resolve(ma);
+        ZipUtils.unzip(ma.getPath(), contentDir);
+    }
+
+    public static void zipServer(Path home, Path contentDir) throws IOException {
+        cleanupServer(home);
+        Path target = contentDir.resolve("wildfly.zip");
+        zip(home, target);
+    }
+
+    private static void cleanupServer(Path jbossHome) throws IOException {
+        Path history = jbossHome.resolve("standalone").resolve("configuration").resolve("standalone_xml_history");
+        IoUtils.recursiveDelete(history);
+        Files.deleteIfExists(jbossHome.resolve("README.txt"));
+    }
+
+    public static void zip(Path contentDir, Path jarFile) throws IOException {
+        try (FileOutputStream fos = new FileOutputStream(jarFile.toFile()); ZipOutputStream zos = new ZipOutputStream(fos)) {
+            Files.walkFileTree(contentDir, EnumSet.of(FileVisitOption.FOLLOW_LINKS), 1,
+                    new SimpleFileVisitor<Path>() {
+                @Override
+                public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs)
+                        throws IOException {
+                    if (!contentDir.equals(dir)) {
+                        zip(dir.toFile(), dir.toFile().getName(), zos);
+                    }
+                    return FileVisitResult.CONTINUE;
+                }
+
+                @Override
+                public FileVisitResult visitFile(Path file, BasicFileAttributes attrs)
+                        throws IOException {
+                    zip(file.toFile(), file.toFile().getName(), zos);
+                    return FileVisitResult.CONTINUE;
+                }
+            });
+        }
+    }
+
+    private static void zip(File fileToZip, String fileName, ZipOutputStream zos) throws IOException {
```

Comment:
Is there a reason here we use the old file? I guess it's because of the way the zip is being created. It could use the `FileSystem` to create the archive though and simply use the new file IO utilities.

Reply:
@jamezp , thank-you to highlight this usage. Actually we have already a support for xipping files that is based on the filesystem. I removed this code and call into this support.